File: 0006.ZigZagConversion/ZigZagConversion.c
char* convert(char* s, int numRows) 
{
	if (numRows == 1) return s;
	int i, j, k = 0, step = 2 * numRows - 2, group, basic;
	int slen = strlen(s);
	char *ret = (char *)malloc(slen + 1);
    group = slen / step;
    if (slen % step) group++;
    
    for (i = 0; i < numRows; i++) {
        for (j = 0; j < group; j++) {
            basic = j * step + i;
            //printf("%d, %d, %d\n", i, j, basic);
            if (basic < slen) ret[k++] = s[basic];
            basic = basic + step - 2*i;
            if (basic < slen && i != 0 && i != numRows - 1) ret[k++] = s[basic];
        }
    }
    ret[k] = '\0';
    return ret;
}

File: 0006.ZigZagConversion/ZigZagConversion.py
class Solution:
    def convert(self, s: str, numRows: int) -> str:
        if numRows == 1: return s
        slen = len(s)
        step = numRows * 2 - 2
        group = slen // step
        last = slen % step
        if last != 0:
            group += 1
        ret = ''

        # deal with each line
        for i in range(numRows):
            # deal with each group
            for j in range(group):
                basic = step * j
                if basic + i < slen:
                    ret += s[basic + i]
                x = basic + step - i
                if x < slen and i != 0 and i != numRows - 1:
                    ret += s[x]
        return ret